R/py_func.R

```diff
@@ -0,0 +1,47 @@
+get_signature <- function(sigs) {
+  sig_names <- names(sigs)
+  signature_strings <- lapply(sig_names, function(k) {
+    if (identical(sigs[[k]], quote(expr =)))
+      # arg without default
+      k
+    else {
+      # arg with default
+      py_value_str <- ifelse(
+        is.character(sigs[[k]]),
+        paste0("'", sigs[[k]], "'"),
+        as.character(r_to_py(eval(sigs[[k]])))) 
+      paste0(k, "=", py_value_str)
+    }
+  })
+  paste(signature_strings, collapse = ", ")
+}
+
+#' Wrap an R function in a Python function with the same signature.
+#' 
+#' This function could wrap an R function in a Python function with
+#' the same signature. Note that if the signature of the R function
+#' contain esoteric Python-incompatible constructs, the signature of 
+#' the returned function will be `function(...)`.
+#' 
+#' @param f An R function
+#' @return A Python function that calls the R function `f` with the same signature.
+#' @export
+py_func <- function(f) {
+  tryCatch({
+    sigs <- formals(f)
+    if (is.null(sigs)) {
+      func_signature <- func_pass_args <- ""
+    } else {
+      func_signature <- get_signature(sigs)
+      func_pass_args <- get_signature(
+        lapply(sigs, function(sig) quote(expr =)))
+    }
+    wrap_fn_util <- py_run_string(sprintf("
+def wrap_fn(f):
+  def fn(%s):
+    return f(%s)
+  return fn
+", func_signature, func_pass_args))
+    wrap_fn_util$wrap_fn(f)
+  }, error = function(e) r_to_py(f))
```

Comment:
I don't think the `tryCatch` with fallback to the default behavior is really helping us here. The user *thinks* they got a complete Python signature but instead they end up getting the default `...` style signature. Why is this better than an error? (I'm presuming that the only reason the user is calling `py_func` is because they *require* the correct signature for use with a library that checks callbacks -- silently giving them an unexpected return value seems bad).

Reply:
Makes sense. I’ll change this